src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/SasTokenCache.java

```diff
@@ -9,72 +9,86 @@
 import org.slf4j.Logger;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.stereotype.Service;
+import reactor.util.function.Tuple2;
+import reactor.util.function.Tuples;
 import uk.gov.hmcts.reform.blobrouter.clients.bulkscanprocessor.BulkScanProcessorClient;
+import uk.gov.hmcts.reform.blobrouter.config.TargetStorageAccount;
 import uk.gov.hmcts.reform.blobrouter.exceptions.InvalidSasTokenException;
 
 import java.time.OffsetDateTime;
 import java.time.ZoneOffset;
 import java.time.temporal.TemporalAccessor;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
 
 import static com.azure.storage.common.implementation.Constants.UrlConstants.SAS_EXPIRY_TIME;
 import static com.azure.storage.common.implementation.StorageImplUtils.parseQueryString;
 import static java.time.temporal.ChronoField.INSTANT_SECONDS;
 import static org.slf4j.LoggerFactory.getLogger;
 
 @Service
-public class BulkScanSasTokenCache {
+public class SasTokenCache {
 
-    private static final Logger logger = getLogger(BulkScanSasTokenCache.class);
-    private final BulkScanProcessorClient bulkScanSasTokenClient;
-    private final long refreshSasBeforeExpiry;
+    private static final Logger logger = getLogger(SasTokenCache.class);
 
-    //key= container name, value = sastoken
-    private static Cache<String, String> tokenCache;
+    // key - storage-account name - value - Map {key= container name, value = sastoken}
+    private Map<TargetStorageAccount, Tuple2<Cache<String, String>, Function<String, String>>> cacheManager;
 
-    public BulkScanSasTokenCache(
+    public SasTokenCache(
         BulkScanProcessorClient bulkScanSasTokenClient,
-        @Value("${bulk-scan-cache.refresh-before-expire-in-sec}") long refreshSasBeforeExpiry
+        @Value("${bulk-scan-cache.refresh-before-expire-in-sec}") long bulkscanSasTokenTtl
     ) {
-        this.bulkScanSasTokenClient = bulkScanSasTokenClient;
-        this.refreshSasBeforeExpiry = refreshSasBeforeExpiry;
-        tokenCache = Caffeine.newBuilder()
-            .expireAfter(new BulkScanSasTokenCacheExpiry())
-            .build();
+        cacheManager = new HashMap<>();
```

Comment:
not thread safe.

Reply:
Discussed in person. I'll raise a new PR